### PR TITLE
Added utility to patch configurations

### PR DESCRIPTION
**Pull request by grastogi23**

### CODE REVIEW

**Comment by grastogi23** [comment]
don't merge it yet... I want to discuss this with you first.

**Comment by grastogi23** [comment]
@anpartha @gcox1 - Hi let me know what you think of this utility. Does it scale and can this be generic enough for handling the conversions. I created this one for verizon.

**Review by chaitanyaavi — APPROVED** [review]

File: python/avi/f5_converter/config_patch.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            as json
+        :return:
+        """
+        cfg_objs = new_cfg.get(obj_type, [])
+        if not cfg_objs:
+            log.warn('Could not apply patch %s: %s as no matching obj found',
+                     obj_type, patch_data)
+            return new_cfg
+        for obj in cfg_objs:
+            obj_name = obj['name']
```
Comment (gcox1) [inline python/avi/f5_converter/config_patch.py]: might fail for objects that do not have name (i.e. SystemConfiguration, ControllerProperties, etc.)

File: python/avi/f5_converter/config_patch.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        tenant = tenant_ref.split('/api/tenant/?name=')[1]
+        obj_name = obj['name']
+        old_obj_ref = '%s:%s' % (tenant, obj['name'])
+        old_obj_refs.append(old_obj_ref)
+        old_obj_ref = obj['name']
+        old_obj_refs.append(old_obj_ref)
+        old_obj_ref = '/api/%s/?name=%s&tenant=%s' % (
+            obj_type.lower(), obj['name'], tenant)
+        old_obj_refs.append(old_obj_ref)
+        obj.update(patch_data['patch'])
```
Comment (gcox1) [inline python/avi/f5_converter/config_patch.py]: Does user mark a 'patch' section in the input data?

File: python/avi/f5_converter/config_patch.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        # as the currently the ref does not have tenant or
+        # cloud information.
+        tenant_ref = obj.get('tenant_ref', '/api/tenant/?name=admin')
+        tenant = tenant_ref.split('/api/tenant/?name=')[1]
+        obj_name = obj['name']
+        old_obj_ref = '%s:%s' % (tenant, obj['name'])
+        old_obj_refs.append(old_obj_ref)
+        old_obj_ref = obj['name']
+        old_obj_refs.append(old_obj_ref)
+        old_obj_ref = '/api/%s/?name=%s&tenant=%s' % (
```
Comment (gcox1) [inline python/avi/f5_converter/config_patch.py]: I think order needs to be /api/%s/?tenant=%s&name=%s

File: python/avi/f5_converter/config_patch.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            obj_type.lower(), obj['name'], tenant)
+        old_obj_refs.append(old_obj_ref)
+        obj.update(patch_data['patch'])
+        new_obj_ref = '%s:%s' % (obj.get('tenant', 'admin'), obj['name'])
+        new_obj_name = obj['name']
+        if ('name' in patch_data['patch']) and (obj_name != new_obj_name):
+            # need to update all the references for this object
+            new_obj_ref = '/api/%s/?name=%s&tenant=%s' % (
```
Comment (gcox1) [inline python/avi/f5_converter/config_patch.py]: same here as line 117

File: python/avi/f5_converter/config_patch.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                obj_type.lower(), obj['name'], tenant)
+            # this is to handle old references could be in multiple formats
+            for old_obj_ref in old_obj_refs:
```
Comment (gcox1) [inline python/avi/f5_converter/config_patch.py]: old_obj_ref will accumulate:
\<name\>
\<tenant\>:\<name\>
/api/\<object\>/?tenant=\<tenant\>&name=\<name\>

new_obj_ref will be:
/api/%s/?tenant=\<tenant\>&name=\<name\>

Comment (gcox1) [reply python/avi/f5_converter/config_patch.py]: in both : and /api formats we have name, tenant-name, and tenant-name-cloud:

I see a TODO item that has adding cloud to combinations, so without cloud:
old_obj_ref and new_obj_ref are missing:
/api/\<object\>/?name=\<name\>

^^The ref would look like this if they were to rename a Tenant

File: python/avi/f5_converter/config_patch.py (hunk 1 above)
Comment (grastogi23) [inline python/avi/f5_converter/config_patch.py]: ok

File: python/avi/f5_converter/config_patch.py (hunk 2 above)
Comment (grastogi23) [reply python/avi/f5_converter/config_patch.py]: yes, updating examples to reflect that

File: python/avi/f5_converter/config_patch.py (hunk 3 above)
Comment (grastogi23) [reply python/avi/f5_converter/config_patch.py]: ok

File: python/avi/f5_converter/config_patch.py (hunk 4 above)
Comment (grastogi23) [reply python/avi/f5_converter/config_patch.py]: ok

File: python/avi/f5_converter/config_patch.py (hunk 5 above)
Comment (grastogi23) [reply python/avi/f5_converter/config_patch.py]: I guess, you mean cloud=<cloud>?

Comment (gcox1) [reply python/avi/f5_converter/config_patch.py]: references for tenant will look like:
/api/tenant/?name=tenant

references for applicationprofile will look like:
/api/applicationprofile/?tenant=tenant&name=name

and references for pool will look like:
/api/pool/?tenant=tenant&name=name&cloud=cloud

so the reference that you are looking for/replacing will be different depending on if the object has a tenant_ref or cloud_ref

**Review by chaitanyaavi — APPROVED** [review]

**Pull request merged**